src/utils/personToPatientIdsExpander.js

```diff
@@ -0,0 +1,114 @@
+const async = require('async');
+const {assertTypeEquals} = require('./assertType');
+const {DatabaseQueryFactory} = require('../dataLayer/databaseQueryFactory');
+
+const patientReferencePrefix = 'Patient/';
+const personReferencePrefix = 'Person/';
+const personProxyPrefix = 'person.';
+const patientReferencePlusPersonProxyPrefix = `${patientReferencePrefix}${personProxyPrefix}`;
+const maximumRecursionDepth = 4;
+
+
+class PersonToPatientIdsExpander {
+    /**
+     * constructor
+     * @param {DatabaseQueryFactory} databaseQueryFactory
+     */
+    constructor(
+        {
+            databaseQueryFactory
+        }
+    ) {
+        /**
+         * @type {DatabaseQueryFactory}
+         */
+        this.databaseQueryFactory = databaseQueryFactory;
+        assertTypeEquals(databaseQueryFactory, DatabaseQueryFactory);
+    }
+
+    /**
+     * replaces patient proxy with actual patient ids
+     * @param {string} base_version
+     * @param {string} id
+     * @param {boolean} includePatientPrefix
+     * @return {Promise<string>}
+     */
+    async getPatientProxyIdsAsync({base_version, id, includePatientPrefix}) {
+        const databaseQueryManager = this.databaseQueryFactory.createQuery({
+            resourceType: 'Person',
+            base_version: base_version
+        });
+        // 1. Get person id from id
+        const personId = id.replace(patientReferencePlusPersonProxyPrefix, '').replace(personProxyPrefix, '');
+        // 2. Get that Person resource from the database
+        let patientIds = await this.getPatientIdsFromPersonAsync(
+            {
+                personId,
+                databaseQueryManager,
+                level: 1
+            }
+        );
+        if (patientIds && patientIds.length > 0) {
+            if (includePatientPrefix) {
+                patientIds = patientIds.map(p => `${patientReferencePrefix}${p}`);
+            }
+            // 4. return a csv of those patient ids (remove duplicates)
+            return Array.from(new Set(patientIds)).join(',');
+        }
+        return id;
+    }
+
+    /**
+     * gets patient ids (recursive) from a person
+     * @param {string} personId
+     * @param {DatabaseQueryManager} databaseQueryManager
+     * @param {number} level
+     * @return {Promise<string[]>}
+     */
+    async getPatientIdsFromPersonAsync({personId, databaseQueryManager, level}) {
+        /**
+         * @type {Person|null}
+         */
+        const person = await databaseQueryManager.findOneAsync(
+            {
+                query: {id: personId},
+                options: {projection: {id: 1, link: 1, _id: 0}}
+            }
```

Comment:
Looks like this recursively queries. Do we have the right indexes and/or caching to make sure it's performant?

Reply:
It is searching on id which is always indexed.